### PR TITLE
Improved ODESolver::solve()

### PR DESCRIPTION
**Pull request by regislebrun**
Now the discretization can be given as an 1D mesh. Very useful when dealing with functions producing fields.

### CODE REVIEW

**Comment by dbarbier** [comment]
-1 for me, unless I am mistaken, you can already write `solve(state, grid.getValues())`, so I do not see the point of adding this method.

**Comment by regislebrun** [comment]
@dbarbier getValues() gives you a Sample, not a Point, so it doesn't work.

**Comment by regislebrun** [comment]
@dbarbier I sent a script implementing one of the use cases from EDF and I had to solve an ODE over a grid. I found it painful to have to manage two grids: one for the resolution of the ODE, one for defining the resulting field, so I added this signature. It cannot be restricted to a RegularGrid object as we may need a non-uniform discretization.

**Comment by dbarbier** [comment]
> @dbarbier getValues() gives you a Sample, not a Point, so it doesn't work.

No, a Point.

> I sent a script implementing one of the use cases from EDF and I had to solve an ODE over a grid. 

So `solve(state,grid.getValues())` does the job.

> It cannot be restricted to a RegularGrid object as we may need a non-uniform discretization.

IMHO it would be much more useful to add a method to convert a Sample into a Point, say `solve(state,grid.getVertices().asPoint())`.

**Comment by regislebrun** [comment]
@dbarbier You are right, I made a confusion with Field::getValues(). So if I use a RegularGrid discretization it works, but not for a general mesh.
I also agree with you on the fact that a method converting a Sample into a Point would be useful (and it almost exists: SampleImplementation::getData()), but this method should not be restricted to 1D sample to be fully useful, so it would be possible for a user to use a mesh of any dimension to solve an ODE, which is most likely an error from its side. The dimension check is lost.

From a user perspective, I am not totally convinced that solve(state, grid) and solve(stat, grid.getVertices().asPoint()) are equivalent in terms of simplicity of use, too.

**Comment by dbarbier** [comment]
If someone is able to use a Mesh as a non-uniform time discretization, she certainly wouldn't mind having to add `.getVertices().asPoint()`.  In practice, one would write
```
# Extract time discretization
timeValues = mesh.getVertices().asPoint()
result = solver.solve(initialState, timeValues)
```
which is IMHO as readable as solve(state,mesh).

**Comment by regislebrun** [comment]
@dbarbier Now the user can do things both ways. I would like to keep the test on the mesh dimension, which is lost if the vertices are simply flatten.

File: python/src/Sample_doc.i.in [hunk 1]
```diff
@@ -1104,6 +1104,31 @@ Examples
...
+>>> import openturns as ot
+>>> ot.RandomGenerator.SetSeed(0)
+>>> sample = ot.Normal(2).getSample(3)
+>>> print(sample)
+    [ X0        X1        ]
+0 : [  0.608202 -1.26617  ]
+1 : [ -0.438266  1.20548  ]
+2 : [ -2.18139   0.350042 ]
+>>> print(sample.asPoint())
+[-0.355007,1.43725,0.810668,0.793156,-0.470526,0.261018]"
```
Comment (dbarbier) [inline python/src/Sample_doc.i.in]: This example looks bogus ;-)

**Comment by dbarbier** [comment]
How about restricting `asPoint()` to 1D samples, and throw an exception otherwise?
Most of the time, we will use it like `sample.getMarginal(index).asPoint()`.

**Pull request merged**